File: content/problems/1960-maximum-product-of-the-length-of-two-palindromic-substrings/solutions/manacher-prefix-suffix-maxima/Solution.kt
class Solution {
    fun maxProduct(s:String):Long{
        val n=s.length;
        val radius=IntArray(n);
        var left=0;
        var right=-1;
        for(i in 0 until n){
            var value=if(i>right)1 else minOf(radius[left+right-i],right-i+1);
            while(i-value>=0&&i+value<n&&s[i-value]==s[i+value])value++;
            radius[i]=value;
            if(i+value-1>right){
                left=i-value+1;
                right=i+value-1
            }
        };
        val end=IntArray(n);
        val start=IntArray(n);
        for(i in 0 until n){
            val length=2*radius[i]-1;
            end[i+radius[i]-1]=maxOf(end[i+radius[i]-1],length);
            start[i-radius[i]+1]=maxOf(start[i-radius[i]+1],length)
        };
        for(i in n-2 downTo 0)end[i]=maxOf(end[i],end[i+1]-2);
        for(i in 1 until n)end[i]=maxOf(end[i],end[i-1]);
        for(i in 1 until n)start[i]=maxOf(start[i],start[i-1]-2);
        for(i in n-2 downTo 0)start[i]=maxOf(start[i],start[i+1]);
        var answer=0L;
        for(i in 0 until n-1)answer=maxOf(answer,end[i].toLong()*start[i+1]);
        return answer
    }
}

File: content/problems/1960-maximum-product-of-the-length-of-two-palindromic-substrings/solutions/manacher-prefix-suffix-maxima/solution.cpp
class Solution {
    public: long long maxProduct(string s){
        int n=s.size();
        vector<int> radius(n);
        for(int i=0,left=0,right=-1;i<n;++i){
            int value=i>right?1:min(radius[left+right-i],right-i+1);
            while(i-value>=0&&i+value<n&&s[i-value]==s[i+value])++value;
            radius[i]=value;
            if(i+value-1>right){
                left=i-value+1;
                right=i+value-1;
            }
        }
        vector<int> ending(n),starting(n);
        for(int i=0;i<n;++i){
            int length=2*radius[i]-1;
            ending[i+radius[i]-1]=max(ending[i+radius[i]-1],length);
            starting[i-radius[i]+1]=max(starting[i-radius[i]+1],length);
        }
        for(int i=n-2;i>=0;--i)ending[i]=max(ending[i],ending[i+1]-2);
        for(int i=1;i<n;++i)ending[i]=max(ending[i],ending[i-1]);
        for(int i=1;i<n;++i)starting[i]=max(starting[i],starting[i-1]-2);
        for(int i=n-2;i>=0;--i)starting[i]=max(starting[i],starting[i+1]);
        long long answer=0;
        for(int i=0;i+1<n;++i)answer=max(answer,1LL*ending[i]*starting[i+1]);
        return answer;
    }
};
